File: packages/expo-contacts/ios/next/records/fields/RelationRecord.swift
import ExpoModulesCore

struct NewRelationRecord: NewRecord {
  @Field var name: String
  @Field var label: String?
}

struct ExistingRelationRecord: ExistingRecord {
  @Field(FieldOption.required) var id: String
  @Field var name:String
  @Field var label: String?
  
  init() {}
  
  init(id: String, name: String, label: String?) {
    self.id = id
    self.name = name
    self.label = label
  }
}

struct PatchRelationRecord: PatchRecord {
  @Field(FieldOption.required) var id: String
  @Field var label: ValueOrUndefined<String?> = .undefined
  @Field var name: ValueOrUndefined<String?> = .undefined
}

File: packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/records/fields/RelationRecord.kt
package expo.modules.contacts.next.records.fields

import expo.modules.contacts.next.records.ExistingRecord
import expo.modules.contacts.next.records.NewRecord
import expo.modules.contacts.next.records.PatchRecord
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Required
import expo.modules.kotlin.types.ValueOrUndefined

sealed interface RelationRecord {
  data class Existing(
    @Required @Field override val id: String,
    @Field val name: String? = null,
    @Field val label: String? = null
  ) : ExistingRecord

  data class New(
    @Field val label: String? = null,
    @Field val name: String? = null
  ) : NewRecord

  class Patch() : PatchRecord {
    @Required @Field
    override lateinit var id: String
    @Field val label: ValueOrUndefined<String?> = ValueOrUndefined.Undefined()
    @Field val name: ValueOrUndefined<String?> = ValueOrUndefined.Undefined()
  }
}
